main/src/ca/uwaterloo/flix/language/ast/SyntaxTree.scala

```diff
@@ -0,0 +1,504 @@
+/*
+ * Copyright 2023 Herluf Baggesen
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package ca.uwaterloo.flix.language.ast
+
+import ca.uwaterloo.flix.language.errors.Parser2Error
+
+
+/**
+ * A Syntax tree representing the __syntax__ in a source file.
+ * The tree is built of nodes that hold a [[TreeKind]] and zero or more children.
+ * Each child is either a [[Token]] or another node.
+ * [[SyntaxTree]] is intentionally kept very flexible in definition, so it can encode faulty syntax.
+ * Beware that this means that it gives few guarantees.
+ * There is no guarantee on the amount or presence of children of a certain kind for instance.
+ */
+object SyntaxTree {
+
+  /**
+   * A node in a [[SyntaxTree]]
+   *
+   * @param kind     The kind of the node.
+   * @param loc      The location that the node spans in the source file.
+   * @param children The children of the node.
+   */
+  case class Tree(kind: TreeKind, var loc: SourceLocation, var children: Array[Child]) {
+    /**
+     * Utility function that computes a textual representation of a [[SyntaxTree.Tree]].
+     * Meant for debugging use.
+     */
+    def toDebugString(nesting: Int = 1): String = {
```

Comment:
Perhaps it would be better to move this _functionality_ out of the definition of the tree and into the phase that uses it. Thats typically what we do.

Reply:
I will move it into the parser then 👍
I is quite handy for debugging, so I was not planning on removing it. If some day someone else is going to work on the parser, I think they would need to write something similar.